tools/snapshot_bot_remote_display/snapshot_bot_remote_display.cc

```diff
@@ -0,0 +1,436 @@
+            BOB_ASSERT(m_TestingBackground.size() == m_Config.getResolution());
+            BOB_ASSERT(m_TestingBackground.type() == CV_8UC3);
+        }
+
+        // Start in training state
+        m_StateMachine.transition(State::Training);
+    }
+
+    //------------------------------------------------------------------------
+    // Public API
+    //------------------------------------------------------------------------
+    bool update()
+    {
+        return m_StateMachine.update();
+    }
+
+private:
+    //------------------------------------------------------------------------
+    // FSM::StateHandler virtuals
+    //------------------------------------------------------------------------
+    virtual bool handleEvent(State state, Event event) override
+    {
+        // If there is a state transition to apply, do so
+        const State snapshotBotState = m_SnapshotBotState.exchange(State::Invalid);
+        if(snapshotBotState != State::Invalid && state != snapshotBotState) {
+            m_StateMachine.transition(snapshotBotState);
+        }
+
+        if(state == State::Training) {
+            if(event == Event::Enter) {
+                // Copy training background over output image
+                if(m_TrainingBackground.empty()) {
+                    m_OutputImage.setTo(cv::Scalar(255, 255, 255));
+                }
+                else {
+                    m_TrainingBackground.copyTo(m_OutputImage);
+                }
+            }
+            else if(event == Event::Update) {
+                // Update live snapshot
+                if(m_LiveNetSource.readFrame(m_ReceiveBuffer)) {
+                    resizeAndProcessImageIntoOutputROI(m_Config.getTrainingLiveRect(), m_ReceiveBuffer);
+                }
+
+                // Update live snapshot
+                if(m_SnapshotNetSource.readFrame(m_ReceiveBuffer)) {
+                    // Convert received image into suitable format for displaying
+                    processImage(m_ReceiveBuffer);
+
+                    // Add image to receive buffer
+                    m_TrainingSnapshots.emplace_back();
+                    m_ReceiveBuffer.copyTo(m_TrainingSnapshots.back());
+
+                    const int numTrainingSnapshots = m_TrainingSnapshots.size();
+                    const int numRectangles = m_Config.getTrainingSnapshotRects().size();
+
+                    const int numSnapshotsToDisplay = std::min(numTrainingSnapshots, numRectangles);
+                    const int startDisplaySnapshot = std::max(0, numTrainingSnapshots - numRectangles);
+
+                    // Loop through available rectangles
+                    for(int i = 0; i < numSnapshotsToDisplay; i++) {
+                        resizeImageIntoOutputROI(m_Config.getTrainingSnapshotRects()[i],
+                                                 m_TrainingSnapshots[startDisplaySnapshot + i]);
+                    }
```

Comment:
What are the rectangles for? Are you cropping training images to different sizes or something?

Reply:
yeah, they seemed the most intuitive way of expressing scaling of snapshots into areas of the screen

Reply:
Ah I see